### ISSUE TITLE
Setting Picker Item key prop causes keys to be coerced to a string regardless of the actual type

### ISSUE THREAD

**Issue by bmingles** [adobe/react-spectrum#6103]
### Provide a general summary of the issue here

The Picker component supports items with numeric id / key props. 

### Item.key unset
If the `Item.key` prop is not explicitly set, keys get derived from the `id` or `key` props on the items array.

e.g.

```jsx
function Example() {
  let items = [
    { id: 1, name: 'Aardvark' },
    { id: 2, name: 'Cat' },
    { id: 3, name: 'Dog' },  ];
    
  let [animalId, setAnimalId] = React.useState(null);

  return (
    <>
      <Picker
        label="Pick an animal"
        items={items}
        onSelectionChange={setAnimalId}
        value={animalId}
      >
        { /* Item.key not explicitly set */
          (item) => <Item>{item.name}</Item>
        }
      </Picker>
      <p>Animal id type: {typeof animalId}</p>
    </>
  );
}
```

The `onSelectionChange` handler will pass a numeric key to `setAnimalId` (consistent with the items data) and `value` + `defaultValue` will work with values of type `number`.

### Item.key explicitly set
If the `Item.key` prop is explicitly set e.g.

```jsx
<Item key={item.id}>{item.name}</Item>}
```

The `onSelectionChange` will pass a string id to `setAnimalId`, and the `value` + `defaultValue` props will require values of type `string` in order to work.

### 🤔 Expected Behavior?

I would expect in the case  of setting `Item.key` prop explicitly:
*  `onSelectionChange` should pass the actual value (not the stringified version) of the original id / key
* 'value`, and `defaultValue` should be able to match the id / key prop on the items instead of requiring them to be string values

### 😯 Current Behavior

When `Item.key` prop is explicitly set
*  `onSelectionChange` passes a stringified value of the original id / key
* 'value`, and `defaultValue` don't match the items id / key prop unless they are stringified

### 🔦 Context

We have a wrapper component that needs to explicitly set the `Item.key` prop based on properties on the `items`. In cases where these props are `numeric` or `boolean` types, this forces the input `value` + `onSelectedValue` change handlers to work with strings which doesn't match the original data.

### 🖥️ Steps to Reproduce

https://codesandbox.io/p/sandbox/modern-shadow-znqpwf?file=%2Fsrc%2FApp.js%3A22%2C25&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu8m4b1r00063b6qt0em6lmr%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu8m4b1r00023b6qa3ncq69w%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu8m4b1r00033b6qnvq8g9lk%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu8m4b1r00053b6q8rg0tzb0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu8m4b1r00023b6qa3ncq69w%2522%253A%257B%2522id%2522%253A%2522clu8m4b1r00023b6qa3ncq69w%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu8m4rnj00513b6q8rkyrj65%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A22%252C%2522startColumn%2522%253A25%252C%2522endLineNumber%2522%253A22%252C%2522endColumn%2522%253A25%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clu8m4rnj00513b6q8rkyrj65%2522%257D%252C%2522clu8m4b1r00053b6q8rg0tzb0%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu8m4b1r00043b6qt6yqaeqs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu8m4b1r00053b6q8rg0tzb0%2522%252C%2522activeTabId%2522%253A%2522clu8m4b1r00043b6qt6yqaeqs%2522%257D%252C%2522clu8m4b1r00033b6qnvq8g9lk%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clu8m4b1r00033b6qnvq8g9lk%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

### Version

3.34.1

### What browsers are you seeing the problem on?

Chrome

### What operating system are you using?

MacOS

**Comment by snowystinger** [comment]
if you change from `key` to `id`, it should work. I'm not entirely sure the reason why `key` doesn't, but I know we're trying not to use it so maybe something was overlooked.

**Comment by snowystinger** [comment]
Yes, @sookmax has the correct analysis. I'm not sure I follow why you must use key and can't change to using id though. Could you explain more?

We've discussed a getKey before, I've forgotten the resolution of that discussion. I'll bring it up with the team to see if anyone recalls.

**Comment by bmingles** [comment]
In my case we have implemented Windowed data where data on items is lazy loaded based on scroll position. The item data looks something like:

```typescript
interface {
   key: Key,
   data?: { realKey: number }
}
```

Where the `key` prop is a stable incrementing key that is always present, but the `data.realKey` is not.

To render this we use:
```jsx
(item) => <Item key={item.data?.realKey ?? item.key}>{...}</Item>
```

This lets us incorporate the real key once it exists but fallback to the stable key that is always present if data isn't loaded for the item. I'm not aware of any way to coalesce props using the implicit keys, so have to be explicit.

**Comment by snowystinger** [comment]
Would you still be able to use id?
```
(item) => <Item id={item.data?.realKey ?? item.key}>{...}</Item>
```

**Comment by bmingles** [comment]
Types don't show that `<Item>` supports an `id` prop.

**Comment by snowystinger** [comment]
That's interesting, I thought it was. It appears to works if you use it, so I'm not sure why it's not in the types. We can look into it.

**Comment by sookmax** [comment]
@bmingles 

In fact, `id` set on `<Item>` doesn't seem to be considered at all in `CollectionBuilder.getKey()`. Only the React `key` prop on `<Item>` (prioritized) and `key` (prioritized) or `id` on the data `item` are considered as candidates of key:
https://github.com/adobe/react-spectrum/blob/1293adf00a11ecac84ec9ab19af7dcea970af2b2/packages/%40react-stately/collections/src/CollectionBuilder.ts#L64-L84

> this may be more than a TypeScript types problem. I don't think the id prop quite solves this use case.

Yeah I agree.

> an optional getKey prop that could be used to derive the key from an item. That could allow users to not have to set the Item.key explicitly and instead control how the component derives the key from the item.

This sounds like a good idea to me too.

**Comment by snowystinger** [comment]
Thanks for the link to that piece of code @sookmax, I had forgotten that it was only key on the item which was used and that the other options came from the data object.

I think you could make use of the key attribute on the data objects themselves though, or id. And just not use the key prop on Item?

```
[
      {fakeKey: 1, key: 11, name: 'Aardvark'},
      {fakeKey: 2, name: 'Kangaroo'},
      {fakeKey: 3, key: 33, name: 'Snake'},
      {fakeKey: 4, name: 'Danni'},
      {fakeKey: 5, key: 55, name: 'Devon'},
      {fakeKey: 6, name: 'Ross'},
      {fakeKey: 7, key: 77, name: 'Puppy'},
      {fakeKey: 8, name: 'Doggo'},
      {fakeKey: 9, key: 99, name: 'Floof'}
    ].map(item => {
      if (item.key === undefined) {
        return {...item, key: item.fakeKey};
      }
      return key;
    })
```

**Comment by bmingles** [comment]
@snowystinger that would probably work but for my use case that would require mapping over the entire list any time an item changes which seems not ideal since our lists could be 10K-100K items.

**Comment by sookmax** [comment]
@snowystinger 

Yeah you're right in that the user can always reorganize (duplicate) their data to fit the constraints (here in our case, to have fields named `id` or `key`). 

I just thought it would be nicer to give the user a way to specify the field to be used as key, without them having to preprocess their data to generate a possibly duplicate field for key.

But at the same time, I'm not sure how we could go about achieving that.
Well roughly maybe would something like this work?
```diff
private getKey(item: CollectionElement<T>, partialNode: PartialNode<T>, state: CollectionBuilderState, parentKey?: Key): Key {
  if (item.key != null) {
    return item.key;
  }

  if (partialNode.type === 'cell' && partialNode.key != null) {
    return `${parentKey}${partialNode.key}`;
  }

  let v = partialNode.value as any;
  if (v != null) {
--  let key = v.key ?? v.id;
++  let key;
++  if (item.props.getKey) {
++    key = item.props.getKey(v);
++  } else {
++    key = v.key ?? v.id;
++  }
    if (key == null) {
      throw new Error('No key found for item');
    }

    return key;
  }

  return parentKey ? `${parentKey}.${partialNode.index}` : `$.${partialNode.index}`;
}
```

So that user can specify their key like:
```tsx
(item) => <Item getKey={(item) => item.data?.realKey ?? item.key}>{...}</Item>
```

**Comment by snowystinger** [comment]
> @snowystinger that would probably work but for my use case that would require mapping over the entire list any time an item changes which seems not ideal since our lists could be 10K-100K items.

Not ideal, but theoretically your items aren't changing that frequently. So if you're running into an issue, you could speed this up by doing your own caching so that you don't have `{...item, key: item.fakeKey}` for every item. Also, this happens in linear time.

@sookmax agreed this may be nice; I'm bringing it up with the team this week, hopefully.

Also of note for anyone coming across this, we do change this in the RAC version to use the id prop instead of the key prop. https://github.com/adobe/react-spectrum/blob/e6923d9473264b972974513dacbcfab5a24d5616/packages/react-aria-components/src/Collection.tsx#L677 apologies for the confusion.

**Comment by qmhc** [comment]
Is there any update?

**Comment by snowystinger** [comment]
No updates of note. It's low priority as it has a workaround. The fastest way to see this change would be to submit a PR for further discussion on implementation.

**Comment by bmingles** [comment]
@snowystinger The workaround does not work for the use case I am reporting. Data can change very frequently in which case the mapping is a problem.
